src/js/node/net.ts

```diff
@@ -210,37 +243,43 @@ const Socket = (function (InternalSocket) {
     static #End(socket) {
       const self = socket.data;
       if (!self) return;
-      self.#ended = true;
-      const queue = self.#readQueue;
-      if (queue.isEmpty()) {
-        if (self.push(null)) {
-          return;
+      // we just reuse the same code but we can push null or enqueue right away
+      Socket.#EmitEndNT(self);
+    }
+    static #EmitEndNT(self, err) {
+      if (!self.#ended) {
+        if (!self.allowHalfOpen) {
+          self.write = writeAfterFIN;
         }
+        self.#ended = true;
+        self.push(null);
       }
-      queue.push(null);
+      // TODO: check how the best way to handle this
+      // if (err) {
+      //   self.destroy(err);
+      // }
     }
-    static #Close(socket) {
+    static #Close(socket, err) {
       const self = socket.data;
       if (!self || self.#closed) return;
       self.#closed = true;
       //socket cannot be used after close
       detachSocket(self);
       if (!self.#ended) {
-        const queue = self.#readQueue;
-        if (queue.isEmpty()) {
-          if (self.push(null)) return;
-        }
-        queue.push(null);
+        // close event can be emitted when we still have data to read from the socket
+        // we will force the close (not allowing half open) and emit the end event after some time
+        setTimeout(Socket.#EmitEndNT, 0, self, err);
```

Comment:
Is this setTimeout intentional?

Reply:
Is there an alternative that doesn't rely on timers?

Reply:
actually I dont think we need the setTimeout at all